### PR TITLE
[decompositions] add decomposition for RNN with packed sequence

### PR DESCRIPTION
**Pull request by samdow**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #91466
* #91465
* __->__ #91281
* #91124
* #91807
* #91123

### CODE REVIEW

File: torch/_decomp/decompositions.py [hunk 1]
```diff
@@ -2245,6 +2373,9 @@ def one_layer_lstm(inp, hidden, params, has_biases, reverse=False):
         hx = hy
         cx = cy
 
+    if reverse:
+        step_output.reverse()
+
```
Comment (zou3519) [inline torch/_decomp/decompositions.py:2378]: A bit confused about what's going on here. Why is this necessary for LSTM? (It isn't in the RNN decomp). If it is necessary, then should this be in the previous PR?

Comment (samdow) [reply torch/_decomp/decompositions.py:2378]: Yeah so what's happening here is that all the decompositions are using `_rnn_helper`

But that was baking in an assumption that this was going to be a tensor. This basically replaces [L2141](https://github.com/pytorch/pytorch/pull/91281/files#diff-d25ad317caf41c3fdf1de3bcfc730102c9e4a3e68639559017ae45cb0bc26911L2141) because I didn't really think ahead to PackedSequences when I was doing this

File: torch/_decomp/decompositions.py [hunk 2]
```diff
@@ -2092,6 +2092,70 @@ def params_hiddens(params, hiddens, i, bidirectional):
...
+def update_hidden_for_packed_reverse(
+    cur_hidden, last_batch_size, batch_size, inp_hidden
+):
+    assert last_batch_size < batch_size
+    return torch.concat(
+        (
+            cur_hidden,
+            inp_hidden.narrow(0, last_batch_size, batch_size - last_batch_size),
+        )
+    )
```
Comment (zou3519) [inline torch/_decomp/decompositions.py:2110]: I don't see these two helper functions being used, you probably meant to replace the things in `one_layer_rnn_data` with it.

File: torch/testing/_internal/common_modules.py [hunk 3]
```diff
@@ -984,6 +985,13 @@ def module_inputs_torch_nn_RNN_GRU(module_info, device, dtype, requires_grad, tr
                 reference_fn=partial(no_batch_dim_reference_rnn_gru, batch_first=b_f),
             )
         )
+        samples.append(
+            ModuleInput(
+                constructor_input=FunctionInput(**cons_args),
+                forward_input=FunctionInput(pack_padded_sequence(make_input((5, 2, 2)), torch.tensor([5, 3]))),
```
Comment (zou3519) [inline torch/testing/_internal/common_modules.py:991]: nit: The code looks correct to me, but can you add something with a slightly larger batch size (than 2)? Maybe 5?

File: test/test_expanded_weights.py [hunk 4]
```diff
@@ -474,6 +475,38 @@ def forward(self, input):
         expected_grads = tuple(expected_grad for expected_grad in expected_grads if expected_grad is not None)
         assert [self.assertEqual(actual, 2 * expected) for (actual, expected) in zip(actual_grads, expected_grads)]
 
+    # getting the expected per sample grads for packed sequences is non-trivial. This just smoke checks that the
+    # outputs are the same and the sum of the per sample gradients is the same as the batched gradient
```
Comment (zou3519) [inline test/test_expanded_weights.py:479]: Why is this non-trivial? I would have expected one could convert the PackedSequence back into a list of sequences and then for each sequence, compute the gradient and finally stack all those gradients together to produce the per-sample-gradients

File: test/test_expanded_weights.py [hunk 5]
```diff
@@ -474,6 +475,38 @@ def forward(self, input):
...
+                actual_grads.append(param.grad_sample.sum(0))
+                del param.grad_sample
+
+            module.zero_grad()
+            # get per sample grads with a for loop
+            expected_res = module(input, *args, **kwargs).data.sum()
+            expected_res.backward()
+            expected_grads = []
+            for param in module.parameters():
+                expected_grads.append(param.grad)
```
Comment (zou3519) [inline test/test_expanded_weights.py:504]: Comment isn't correct, right? we're not computing per-sample grads here

File: test/test_expanded_weights.py [hunk 6]
```diff
@@ -474,6 +475,38 @@ def forward(self, input):
...
+        kwargs = kwargs if kwargs is not None else {}
+
+        batch_size = max(tuple(input.batch_sizes)).item()
+
+        with freeze_rng_state():
+            # get per sample grads with ExpandedWeights context manager
+            actual_res = call_for_per_sample_grads(module,
+                                                   batch_size=batch_size,
+                                                   loss_reduction="sum")(input, *args, **kwargs).data.sum()
+            actual_res.backward(retain_graph=True)
```
Comment (zou3519) [inline test/test_expanded_weights.py:491]: nit: don't need retain_graph since we don't backward through actual-res again, right? Unless I'm misreading

File: torch/nn/utils/_expanded_weights/expanded_weights_utils.py [hunk 7]
```diff
@@ -90,6 +91,15 @@ def set_grad_sample_if_exists(maybe_expanded_weight, per_sample_grad_fn):
...
     if isinstance(maybe_expanded_weight, ExpandedWeight):
         grad_sample_contribution = maybe_scale_by_batch_size(per_sample_grad_fn(unpacked), maybe_expanded_weight)
+
+        if maybe_expanded_weight.batch_size > grad_sample_contribution.shape[0]:
+            # this only passes the other checks if the arg allows smaller batch sizes
+            int = torch.zeros(maybe_expanded_weight.batch_size, *grad_sample_contribution.shape[1:],
+                              dtype=grad_sample_contribution.dtype,
+                              device=grad_sample_contribution.device)
+            int[:grad_sample_contribution.shape[0]] = grad_sample_contribution
+            grad_sample_contribution = int
```
Comment (zou3519) [inline torch/nn/utils/_expanded_weights/expanded_weights_utils.py:101]: nit: `int` is special in Python, let's rename it and not shadow

**Review by zou3519 — APPROVED** [review]
LGTM, PackedSequence is wild. Would prefer some more test cases (e.g. batch_size > 2). Also It would be good to actually compute the per sample grads and do the comparison instead of the smoke test (I'm not sure how much work this is, if you don't have bandwidth then please file an issue and we'll get to it)

File: torch/_decomp/decompositions.py (hunk 1 above)
Comment (zou3519) [reply torch/_decomp/decompositions.py:2378]: Thanks for pointing that out, I missed that line

File: torch/_decomp/decompositions.py (hunk 2 above)
Comment (samdow) [reply torch/_decomp/decompositions.py]: Good catch! Fixed so that they're being used

File: test/test_expanded_weights.py (hunk 4 above)
Comment (samdow) [reply test/test_expanded_weights.py]: Good point! I got confused with how packed sequence is implemented and forgot about the inverse operation. Added the test but the one that's a little tough now is the gradient with respect to the input since they're now taking in two different types so I left out that check

File: torch/nn/utils/_expanded_weights/expanded_weights_utils.py (hunk 7 above)
Comment (samdow) [reply torch/nn/utils/_expanded_weights/expanded_weights_utils.py]: Done!

File: test/test_expanded_weights.py (hunk 6 above)
Comment (samdow) [reply test/test_expanded_weights.py]: Yeah great question I think I just was trying to make it work instead of actually looking into what's happening. Basically with packed sequences, this got weird since I was setting requires_grad on the tensor used to make the packed sequence. I wasn't able to unset the grad of that tensor since I only had access to the PackedSequence (which only has a tensor that's the output from a function, not the original tensor)

**Comment by ezyang** [comment]
We need someone else to shepherd this right
